lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java

```diff
@@ -19,15 +19,363 @@
+          || sc.getAnswerableMeasureIndices().contains(phrase.getPosition())) {
+        ASTNode node = getSelectExprAST();
+        node.addChild(getInnerSelectExprForAlias(sc, aliasNode.getText()));
+        node.addChild(aliasNode);
+        innerSelect.addChild(node);
+      } else {
+        innerSelect.addChild(getDefaultNode(aliasNode));
+      }
+    }
+    sc.getQueryAst().setSelectAST(innerSelect);
+  }
+
+
+  private ASTNode getInnerSelectExprForAlias(StorageCandidate sc, String alias) {
+    ASTNode expr = new ASTNode();
+    ASTNode select = sc.getQueryAst().getSelectAST();
+    for (int i = 0; i < select.getChildCount(); i++) {
+      String innerAlias = select.getChild(i).getChild(1).getText();
+      if (innerAlias.equals(alias)) {
+        expr = (ASTNode) select.getChild(i).getChild(0);
+        break;
+      }
+    }
+    return expr;
+  }
+
+  private void processSelectAndHavingAST() throws LensException {
+    ASTNode outerSelectAst = new ASTNode(queryAst.getSelectAST());
+    ASTNode outerHavingAst = new ASTNode(new CommonToken(TOK_HAVING, "TOK_HAVING"));
+    for (StorageCandidate sc : storageCandidates) {
+      ASTNode innerSelectAST = new ASTNode(new CommonToken(TOK_SELECT, "TOK_SELECT"));
+      innerToOuterASTs.clear();
+      AliasDecider aliasDecider = new DefaultAliasDecider();
+      ASTNode originalSelectAST = MetastoreUtil.copyAST(sc.getQueryAst().getSelectAST());
+      queryAst.setSelectAST(new ASTNode(originalSelectAST.getToken()));
+      processSelectExpression(sc, outerSelectAst, innerSelectAST, aliasDecider);
+      outerHavingAst = processHavingAST(innerSelectAST, aliasDecider, sc);
+    }
+    queryAst.setSelectAST(outerSelectAst);
+    queryAst.setHavingAST(outerHavingAst);
+  }
+
+  private void processSelectExpression(StorageCandidate sc, ASTNode outerSelectAst,
+                                       ASTNode innerSelectAST, AliasDecider aliasDecider) throws LensException {
+    ASTNode selectAST = sc.getQueryAst().getSelectAST();
+    if (selectAST == null) {
+      return;
+    }
+    // iterate over all children of the ast and get outer ast corresponding to it.
+    for (int i = 0; i < selectAST.getChildCount(); i++) {
+      //for (Node node : selectAST.getChildren()) {
+      ASTNode child = (ASTNode) selectAST.getChild(i);
+      ASTNode outerSelect = new ASTNode(child);
+      ASTNode selectExprAST = (ASTNode) child.getChild(0);
+      ASTNode outerAST = getOuterAST(selectExprAST, innerSelectAST, aliasDecider, sc);
+      outerSelect.addChild(outerAST);
+      // has an alias? add it
+      if (child.getChildCount() > 1) {
+        outerSelect.addChild(child.getChild(1));
+      }
+      if (outerSelectAst.getChildCount() <= selectAST.getChildCount()) {
+        if (outerSelectAst.getChild(i) == null) {
+          outerSelectAst.addChild(outerSelect);
+        } else if (HQLParser.getString((ASTNode) outerSelectAst.getChild(i).getChild(0)).equals("0.0")) {
```

Comment:
what about the case where its not 0.0 ? Should we not update the outer ast in this case with new alias  .. sum(alias1) as Total Burn ?